File: runtime/graphics/csharp/GraphicsRuntimeSDL2.cs
// GraphicsRuntime.cs - SDL2 runtime for goany graphics package
// Requires: SDL2-CS NuGet package

using System;
using SDL2;

public static class graphics
{
    public struct Color
    {
        public byte R;
        public byte G;
        public byte B;
        public byte A;

        public Color(byte r, byte g, byte b, byte a)
        {
            R = r; G = g; B = b; A = a;
        }
    }

    public struct Rect
    {
        public int X;
        public int Y;
        public int Width;
        public int Height;

        public Rect(int x, int y, int width, int height)
        {
            X = x; Y = y; Width = width; Height = height;
        }
    }

    public struct Window
    {
        public long handle;   // SDL_Window*
        public long renderer; // SDL_Renderer*
        public int width;
        public int height;
        public bool running;
    }

    // Global to store last key pressed
    private static int lastKeyPressed = 0;

    // --- Color constructors ---

    public static Color NewColor(byte r, byte g, byte b, byte a)
    {
        return new Color(r, g, b, a);
    }

    public static Color Black() { return new Color(0, 0, 0, 255); }
    public static Color White() { return new Color(255, 255, 255, 255); }
    public static Color Red() { return new Color(255, 0, 0, 255); }
    public static Color Green() { return new Color(0, 255, 0, 255); }
    public static Color Blue() { return new Color(0, 0, 255, 255); }

    // --- Rect constructor ---

    public static Rect NewRect(int x, int y, int width, int height)
    {
        return new Rect(x, y, width, height);
    }

    // --- Window management ---

    public static Window CreateWindow(string title, int width, int height)
    {
        if (SDL.SDL_Init(SDL.SDL_INIT_VIDEO) < 0)
        {
            return new Window { handle = 0, renderer = 0, width = width, height = height, running = false };
        }

        IntPtr sdlWindow = SDL.SDL_CreateWindow(
            title,
            SDL.SDL_WINDOWPOS_CENTERED,
            SDL.SDL_WINDOWPOS_CENTERED,
            width,
            height,
            SDL.SDL_WindowFlags.SDL_WINDOW_SHOWN
        );

        if (sdlWindow == IntPtr.Zero)
        {
            return new Window { handle = 0, renderer = 0, width = width, height = height, running = false };
        }

        IntPtr sdlRenderer = SDL.SDL_CreateRenderer(
            sdlWindow, -1,
            SDL.SDL_RendererFlags.SDL_RENDERER_ACCELERATED | SDL.SDL_RendererFlags.SDL_RENDERER_PRESENTVSYNC
        );

        if (sdlRenderer == IntPtr.Zero)
        {
            SDL.SDL_DestroyWindow(sdlWindow);
            return new Window { handle = 0, renderer = 0, width = width, height = height, running = false };
        }

        return new Window
        {
            handle = sdlWindow.ToInt64(),
            renderer = sdlRenderer.ToInt64(),
            width = width,
            height = height,
            running = true
        };
    }

    public static void CloseWindow(Window w)
    {
        if (w.renderer != 0)
        {
            SDL.SDL_DestroyRenderer(new IntPtr(w.renderer));
        }
        if (w.handle != 0)
        {
            SDL.SDL_DestroyWindow(new IntPtr(w.handle));
        }
        SDL.SDL_Quit();
    }

    public static bool IsRunning(Window w)
    {
        return w.running;
    }

    public static (Window, bool) PollEvents(Window w)
    {
        lastKeyPressed = 0;
        SDL.SDL_Event e;
        while (SDL.SDL_PollEvent(out e) != 0)
        {
            if (e.type == SDL.SDL_EventType.SDL_QUIT)
            {
                w.running = false;
                return (w, false);
            }
            if (e.type == SDL.SDL_EventType.SDL_KEYDOWN)
            {
                SDL.SDL_Keycode key = e.key.keysym.sym;
                // Convert SDL keycode to ASCII for printable characters
                if (key >= SDL.SDL_Keycode.SDLK_SPACE && key <= SDL.SDL_Keycode.SDLK_z)
                {
                    // Check for shift modifier for uppercase
                    SDL.SDL_Keymod mod = (SDL.SDL_Keymod)e.key.keysym.mod;
                    if ((mod & (SDL.SDL_Keymod.KMOD_LSHIFT | SDL.SDL_Keymod.KMOD_RSHIFT)) != 0)
                    {
                        if (key >= SDL.SDL_Keycode.SDLK_a && key <= SDL.SDL_Keycode.SDLK_z)
                        {
                            lastKeyPressed = (int)key - 32; // Convert to uppercase
                        }
                        else
                        {
                            lastKeyPressed = (int)key;
                        }
                    }
                    else
                    {
                        lastKeyPressed = (int)key;
                    }
                }
                else if (key == SDL.SDL_Keycode.SDLK_RETURN)
                {
                    lastKeyPressed = 13; // Enter
                }
                else if (key == SDL.SDL_Keycode.SDLK_BACKSPACE)
                {
                    lastKeyPressed = 8; // Backspace
                }
            }
        }
        return (w, true);
    }

    public static int GetLastKey()
    {
        return lastKeyPressed;
    }

    // GetMouse returns mouse position and button state
    public static (int, int, int) GetMouse(Window w)
    {
        int x, y;
        uint buttons = SDL.SDL_GetMouseState(out x, out y);
        int result = 0;
        if ((buttons & SDL.SDL_BUTTON(SDL.SDL_BUTTON_LEFT)) != 0) result |= 1;
        if ((buttons & SDL.SDL_BUTTON(SDL.SDL_BUTTON_RIGHT)) != 0) result |= 2;
        if ((buttons & SDL.SDL_BUTTON(SDL.SDL_BUTTON_MIDDLE)) != 0) result |= 4;
        return (x, y, result);
    }

    public static int GetWidth(Window w) { return w.width; }
    public static int GetHeight(Window w) { return w.height; }

    // GetScreenSize returns screen resolution (SDL2 cross-platform)
    public static (int, int) GetScreenSize()
    {
        SDL.SDL_DisplayMode mode;
        if (SDL.SDL_GetCurrentDisplayMode(0, out mode) == 0)
        {
            return (mode.w, mode.h);
        }
        return (1920, 1080);
    }

    // CreateWindowFullscreen creates a fullscreen window
    public static Window CreateWindowFullscreen(string title, int width, int height)
    {
        if (SDL.SDL_Init(SDL.SDL_INIT_VIDEO) < 0)
        {
            return new Window { handle = 0, renderer = 0, width = width, height = height, running = false };
        }

        IntPtr window = SDL.SDL_CreateWindow(title,
            SDL.SDL_WINDOWPOS_CENTERED, SDL.SDL_WINDOWPOS_CENTERED,
            width, height, SDL.SDL_WindowFlags.SDL_WINDOW_SHOWN | SDL.SDL_WindowFlags.SDL_WINDOW_FULLSCREEN_DESKTOP);
        if (window == IntPtr.Zero)
        {
            return new Window { handle = 0, renderer = 0, width = width, height = height, running = false };
        }

        IntPtr renderer = SDL.SDL_CreateRenderer(window, -1,
            SDL.SDL_RendererFlags.SDL_RENDERER_ACCELERATED | SDL.SDL_RendererFlags.SDL_RENDERER_PRESENTVSYNC);
        if (renderer == IntPtr.Zero)
        {
            SDL.SDL_DestroyWindow(window);
            return new Window { handle = 0, renderer = 0, width = width, height = height, running = false };
        }

        return new Window
        {
            handle = window.ToInt64(),
            renderer = renderer.ToInt64(),
            width = width,
            height = height,
            running = true
        };
    }

    // --- Rendering ---

    public static void Clear(Window w, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);
        SDL.SDL_RenderClear(renderer);
    }

    public static void Present(Window w)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_RenderPresent(renderer);
    }

    // --- Drawing primitives ---

    public static void DrawRect(Window w, Rect rect, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);
        SDL.SDL_Rect sdlRect = new SDL.SDL_Rect { x = rect.X, y = rect.Y, w = rect.Width, h = rect.Height };
        SDL.SDL_RenderDrawRect(renderer, ref sdlRect);
    }

    public static void FillRect(Window w, Rect rect, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);
        SDL.SDL_Rect sdlRect = new SDL.SDL_Rect { x = rect.X, y = rect.Y, w = rect.Width, h = rect.Height };
        SDL.SDL_RenderFillRect(renderer, ref sdlRect);
    }

    public static void DrawLine(Window w, int x1, int y1, int x2, int y2, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);
        SDL.SDL_RenderDrawLine(renderer, x1, y1, x2, y2);
    }

    public static void DrawPoint(Window w, int x, int y, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);
        SDL.SDL_RenderDrawPoint(renderer, x, y);
    }

    public static void DrawCircle(Window w, int centerX, int centerY, int radius, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);

        // Bresenham's circle algorithm
        int x = radius;
        int y = 0;
        int err = 0;

        while (x >= y)
        {
            SDL.SDL_RenderDrawPoint(renderer, centerX + x, centerY + y);
            SDL.SDL_RenderDrawPoint(renderer, centerX + y, centerY + x);
            SDL.SDL_RenderDrawPoint(renderer, centerX - y, centerY + x);
            SDL.SDL_RenderDrawPoint(renderer, centerX - x, centerY + y);
            SDL.SDL_RenderDrawPoint(renderer, centerX - x, centerY - y);
            SDL.SDL_RenderDrawPoint(renderer, centerX - y, centerY - x);
            SDL.SDL_RenderDrawPoint(renderer, centerX + y, centerY - x);
            SDL.SDL_RenderDrawPoint(renderer, centerX + x, centerY - y);

            y++;
            err += 1 + 2 * y;
            if (2 * (err - x) + 1 > 0)
            {
                x--;
                err += 1 - 2 * x;
            }
        }
    }

    public static void FillCircle(Window w, int centerX, int centerY, int radius, Color c)
    {
        IntPtr renderer = new IntPtr(w.renderer);
        SDL.SDL_SetRenderDrawColor(renderer, c.R, c.G, c.B, c.A);

        for (int y = -radius; y <= radius; y++)
        {
            for (int x = -radius; x <= radius; x++)
            {
                if (x * x + y * y <= radius * radius)
                {
                    SDL.SDL_RenderDrawPoint(renderer, centerX + x, centerY + y);
                }
            }
        }
    }

    // RunLoop runs the main loop, calling frameFunc each frame.
    // frameFunc receives the window and returns true to continue, false to stop.
    // This is the preferred way to write cross-platform graphics code that works in browsers.
    public static void RunLoop(Window w, Func<Window, bool> frameFunc)
    {
        while (true)
        {
            bool running;
            (w, running) = PollEvents(w);
            if (!running)
            {
                break;
            }
            if (!frameFunc(w))
            {
                break;
            }
        }
    }

    // RunLoopWithState runs the main loop with explicit state threading.
    // The state is passed to frameFunc and returned each frame, avoiding closure capture.
    // frameFunc receives the window and state, returns updated state and true to continue.
    public static void RunLoopWithState<S>(Window w, S state, Func<Window, S, (S, bool)> frameFunc)
    {
        while (true)
        {
            bool running;
            (w, running) = PollEvents(w);
            if (!running)
            {
                break;
            }
            bool cont;
            (state, cont) = frameFunc(w, state);
            if (!cont)
            {
                break;
            }
        }
    }
}

File: runtime/graphics/rust/graphics_runtime_sdl2.rs
// graphics_runtime.rs - SDL2 runtime for goany graphics package
// Requires: sdl2 crate in Cargo.toml

use sdl2::event::Event;
use sdl2::keyboard::Keycode;
use sdl2::pixels::Color as SdlColor;
use sdl2::rect::Rect as SdlRect;
use sdl2::render::Canvas;
use sdl2::video::Window as SdlWindow;
use sdl2::Sdl;
use std::cell::RefCell;
use std::collections::HashMap;

// Thread-local storage for SDL context and canvases
thread_local! {
    static SDL_CONTEXT: RefCell<Option<Sdl>> = RefCell::new(None);
    static CANVASES: RefCell<HashMap<i64, Canvas<SdlWindow>>> = RefCell::new(HashMap::new());
    static NEXT_HANDLE: RefCell<i64> = RefCell::new(1);
    static LAST_KEY: RefCell<i32> = RefCell::new(0);
}

#[derive(Clone, Copy, Default, Debug, Hash, PartialEq, Eq)]
pub struct Color {
    pub R: u8,
    pub G: u8,
    pub B: u8,
    pub A: u8,
}

#[derive(Clone, Copy, Default, Debug, Hash, PartialEq, Eq)]
pub struct Rect {
    pub X: i32,
    pub Y: i32,
    pub Width: i32,
    pub Height: i32,
}

#[derive(Clone, Copy, Default, Debug, Hash, PartialEq, Eq)]
pub struct Window {
    pub handle: i64,
    pub renderer: i64,
    pub width: i32,
    pub height: i32,
    pub running: bool,
}

// --- Color constructors ---

pub fn NewColor(r: u8, g: u8, b: u8, a: u8) -> Color {
    Color { R: r, G: g, B: b, A: a }
}

pub fn Black() -> Color { Color { R: 0, G: 0, B: 0, A: 255 } }
pub fn White() -> Color { Color { R: 255, G: 255, B: 255, A: 255 } }
pub fn Red() -> Color { Color { R: 255, G: 0, B: 0, A: 255 } }
pub fn Green() -> Color { Color { R: 0, G: 255, B: 0, A: 255 } }
pub fn Blue() -> Color { Color { R: 0, G: 0, B: 255, A: 255 } }

// --- Rect constructor ---

pub fn NewRect(x: i32, y: i32, width: i32, height: i32) -> Rect {
    Rect { X: x, Y: y, Width: width, Height: height }
}

// --- Window management ---

pub fn CreateWindow(title: String, width: i32, height: i32) -> Window {
    SDL_CONTEXT.with(|ctx| {
        let mut ctx_ref = ctx.borrow_mut();
        if ctx_ref.is_none() {
            *ctx_ref = sdl2::init().ok();
        }

        if let Some(ref sdl) = *ctx_ref {
            let video = match sdl.video() {
                Ok(v) => v,
                Err(_) => return Window { handle: 0, renderer: 0, width, height, running: false },
            };

            let window = match video
                .window(&title, width as u32, height as u32)
                .position_centered()
                .build()
            {
                Ok(w) => w,
                Err(_) => return Window { handle: 0, renderer: 0, width, height, running: false },
            };

            let canvas = match window.into_canvas().accelerated().present_vsync().build() {
                Ok(c) => c,
                Err(_) => return Window { handle: 0, renderer: 0, width, height, running: false },
            };

            let handle = NEXT_HANDLE.with(|h| {
                let mut h_ref = h.borrow_mut();
                let current = *h_ref;
                *h_ref += 1;
                current
            });

            CANVASES.with(|c| {
                c.borrow_mut().insert(handle, canvas);
            });

            Window {
                handle,
                renderer: handle,
                width,
                height,
                running: true,
            }
        } else {
            Window { handle: 0, renderer: 0, width, height, running: false }
        }
    })
}

pub fn CloseWindow(mut w: Window) {
    CANVASES.with(|c| {
        c.borrow_mut().remove(&w.handle);
    });
}

pub fn IsRunning(mut w: Window) -> bool {
    w.running
}

pub fn PollEvents(mut w: Window) -> (Window, bool) {
    LAST_KEY.with(|k| *k.borrow_mut() = 0);
    SDL_CONTEXT.with(|ctx| {
        if let Some(ref sdl) = *ctx.borrow() {
            let mut event_pump = match sdl.event_pump() {
                Ok(ep) => ep,
                Err(_) => return (w, true),
            };

            for event in event_pump.poll_iter() {
                match event {
                    Event::Quit { .. } => {
                        w.running = false;
                        return (w, false);
                    }
                    Event::KeyDown { keycode: Some(key), keymod, .. } => {
                        let ascii = match key {
                            Keycode::Return => 13,
                            Keycode::Backspace => 8,
                            Keycode::Space => 32,
                            k if k as i32 >= Keycode::A as i32 && k as i32 <= Keycode::Z as i32 => {
                                let base = k as i32 - Keycode::A as i32;
                                if keymod.contains(sdl2::keyboard::Mod::LSHIFTMOD) ||
                                   keymod.contains(sdl2::keyboard::Mod::RSHIFTMOD) {
                                    65 + base // Uppercase A-Z
                                } else {
                                    97 + base // Lowercase a-z
                                }
                            }
                            k if k as i32 >= Keycode::Num0 as i32 && k as i32 <= Keycode::Num9 as i32 => {
                                48 + (k as i32 - Keycode::Num0 as i32)
                            }
                            _ => 0,
                        };
                        if ascii != 0 {
                            LAST_KEY.with(|k| *k.borrow_mut() = ascii);
                        }
                    }
                    _ => {}
                }
            }
        }
        (w, true)
    })
}

pub fn GetLastKey() -> i32 {
    LAST_KEY.with(|k| *k.borrow())
}

/// GetMouse returns mouse position and button state
pub fn GetMouse(w: Window) -> (i32, i32, i32) {
    SDL_CONTEXT.with(|ctx| {
        if let Some(ref sdl) = *ctx.borrow() {
            if let Ok(event_pump) = sdl.event_pump() {
                let state = event_pump.mouse_state();
                let mut buttons: i32 = 0;
                if state.left() { buttons |= 1; }
                if state.right() { buttons |= 2; }
                if state.middle() { buttons |= 4; }
                return (state.x(), state.y(), buttons);
            }
        }
        (0, 0, 0)
    })
}

pub fn GetWidth(mut w: Window) -> i32 { w.width }
pub fn GetHeight(mut w: Window) -> i32 { w.height }

/// GetScreenSize returns the screen resolution (SDL2 cross-platform)
pub fn GetScreenSize() -> (i32, i32) {
    SDL_CONTEXT.with(|ctx| {
        let mut ctx_ref = ctx.borrow_mut();
        if ctx_ref.is_none() {
            *ctx_ref = sdl2::init().ok();
        }
        if let Some(ref sdl) = *ctx_ref {
            if let Ok(video) = sdl.video() {
                if let Ok(mode) = video.current_display_mode(0) {
                    return (mode.w, mode.h);
                }
            }
        }
        (1920, 1080)
    })
}

/// CreateWindowFullscreen creates a fullscreen window
pub fn CreateWindowFullscreen(title: String, width: i32, height: i32) -> Window {
    SDL_CONTEXT.with(|ctx| {
        let mut ctx_ref = ctx.borrow_mut();
        if ctx_ref.is_none() {
            *ctx_ref = sdl2::init().ok();
        }

        if let Some(ref sdl) = *ctx_ref {
            let video = match sdl.video() {
                Ok(v) => v,
                Err(_) => return Window { handle: 0, renderer: 0, width, height, running: false },
            };

            let window = match video
                .window(&title, width as u32, height as u32)
                .position_centered()
                .fullscreen_desktop()
                .build()
            {
                Ok(w) => w,
                Err(_) => return Window { handle: 0, renderer: 0, width, height, running: false },
            };

            let canvas = match window.into_canvas().accelerated().present_vsync().build() {
                Ok(c) => c,
                Err(_) => return Window { handle: 0, renderer: 0, width, height, running: false },
            };

            let handle = NEXT_HANDLE.with(|h| {
                let mut h_ref = h.borrow_mut();
                let current = *h_ref;
                *h_ref += 1;
                current
            });

            CANVASES.with(|c| {
                c.borrow_mut().insert(handle, canvas);
            });

            return Window {
                handle,
                renderer: handle,
                width,
                height,
                running: true,
            };
        }
        Window { handle: 0, renderer: 0, width, height, running: false }
    })
}

// --- Rendering ---

pub fn Clear(mut w: Window, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));
            canvas.clear();
        }
    });
}

pub fn Present(mut w: Window) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.present();
        }
    });
}

// --- Drawing primitives ---

pub fn DrawRect(mut w: Window, mut rect: Rect, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));
            let _ = canvas.draw_rect(SdlRect::new(rect.X, rect.Y, rect.Width as u32, rect.Height as u32));
        }
    });
}

pub fn FillRect(mut w: Window, mut rect: Rect, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));
            let _ = canvas.fill_rect(SdlRect::new(rect.X, rect.Y, rect.Width as u32, rect.Height as u32));
        }
    });
}

pub fn DrawLine(mut w: Window, x1: i32, y1: i32, x2: i32, y2: i32, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));
            let _ = canvas.draw_line((x1, y1), (x2, y2));
        }
    });
}

pub fn DrawPoint(mut w: Window, x: i32, y: i32, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));
            let _ = canvas.draw_point((x, y));
        }
    });
}

pub fn DrawCircle(mut w: Window, centerX: i32, centerY: i32, radius: i32, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));

            // Bresenham's circle algorithm
            let mut x = radius;
            let mut y = 0;
            let mut err = 0;

            while x >= y {
                let _ = canvas.draw_point((centerX + x, centerY + y));
                let _ = canvas.draw_point((centerX + y, centerY + x));
                let _ = canvas.draw_point((centerX - y, centerY + x));
                let _ = canvas.draw_point((centerX - x, centerY + y));
                let _ = canvas.draw_point((centerX - x, centerY - y));
                let _ = canvas.draw_point((centerX - y, centerY - x));
                let _ = canvas.draw_point((centerX + y, centerY - x));
                let _ = canvas.draw_point((centerX + x, centerY - y));

                y += 1;
                err += 1 + 2 * y;
                if 2 * (err - x) + 1 > 0 {
                    x -= 1;
                    err += 1 - 2 * x;
                }
            }
        }
    });
}

pub fn FillCircle(mut w: Window, centerX: i32, centerY: i32, radius: i32, mut c: Color) {
    CANVASES.with(|canvases| {
        if let Some(canvas) = canvases.borrow_mut().get_mut(&w.handle) {
            canvas.set_draw_color(SdlColor::RGBA(c.R, c.G, c.B, c.A));

            for y in -radius..=radius {
                for x in -radius..=radius {
                    if x * x + y * y <= radius * radius {
                        let _ = canvas.draw_point((centerX + x, centerY + y));
                    }
                }
            }
        }
    });
}

/// RunLoop runs the main loop, calling frameFunc each frame.
/// frameFunc receives the window and returns true to continue, false to stop.
/// This is the preferred way to write cross-platform graphics code that works in browsers.
pub fn RunLoop<F>(mut w: Window, mut frameFunc: F)
where
    F: FnMut(Window) -> bool,
{
    loop {
        let running: bool;
        (w, running) = PollEvents(w);
        if !running {
            break;
        }
        if !frameFunc(w) {
            break;
        }
    }
}

/// RunLoopWithState runs the main loop with explicit state threading.
/// The state is passed to frameFunc and returned each frame, avoiding closure capture.
/// This eliminates the need to clone state in FnMut closures.
/// frameFunc receives the window and state, returns updated state and true to continue.
pub fn RunLoopWithState<S, F>(mut w: Window, mut state: S, mut frameFunc: F)
where
    F: FnMut(Window, S) -> (S, bool),
{
    loop {
        let running: bool;
        (w, running) = PollEvents(w);
        if !running {
            break;
        }
        let cont: bool;
        (state, cont) = frameFunc(w, state);
        if !cont {
            break;
        }
    }
}
